feat: add database analyse-gas-usage command

jancionear opened the pull request:
Add a command which allows to analyse gas usage on some part of the blockchain.
The command goes over the specified blocks and gathers information about how much
gas was used by each block, shard and account.
Then it displays an analysis of all the data.

The analysis contains:
* Total gas used
* Gas used on each shard
* Top 10 accounts by gas usage
* The optimal account by which a shard could be split into two halves with similar gas usage

Help page:
```bash
$ ./neard database analyse-gas-usage --help
Analyse gas usage in a chosen sequnce of blocks

Usage: neard database analyse-gas-usage [OPTIONS]

Options:
      --last-blocks <LAST_BLOCKS>
          Analyse the last N blocks in the blockchain
      --from-block-height <FROM_BLOCK_HEIGHT>
          Analyse blocks from the given block height, inclusive
      --to-block-height <TO_BLOCK_HEIGHT>
          Analyse blocks up to the given block height, inclusive
  -h, --help
          Print help
```

Example usage:
```bash
# Analyse gas usage in the last 1000 blocks
cargo run --bin neard -- database analyse-gas-usage --last-blocks 1000

# Analyse gas usage in blocks with heights between 200 and 300 (inclusive)
./neard database analyse-gas-usage --from-block-height 200 --to-block-height 300
```

jancionear commented:
Notes:
* I'm not entirely sure that the technique used in `get_gas_usage_in_block` is the correct way to measure used gas. Gas is a tricky subject, I would appreciate if someone with more knowledge about the subject would chime in.
* There's a lot of `unwraps` in the code. Some people dislike that, but in my opinion `unwrap()` is a good choice for fatal errors in user facing applications. An `unwrap()` prints the whole error and the exact line where the error has occured, which is great for finding out what went wrong. With `Result` I often ended up with messages like: `Failed: Error: NotFound`, which isn't very useful.
* The tool isn't very fast - analysing a whole epoch (43200 blocks) takes about 8 minutes. This is most likely caused by the fact that it reads one piece of information from the database at a time. If needed it could probably be sped up by running multiple `get_gas_usage_in_block` in parallel.
* Everything fits in the memory. The tool only keeps one u64 per account. Analysing a whole epoch of blocks required only ~300MB of RAM.

jancionear commented:
Example output:
```bash
$ time ./target/release/neard database analyse-gas-usage --last-blocks 43200
2023-11-23T16:39:01.093282Z  INFO neard: version="trunk" build="1.1.0-4597-gb7d470924" latest_protocol=64
2023-11-23T16:39:01.093504Z  INFO config: Validating Config, extracted from config.json...
2023-11-23T16:39:01.095173Z  WARN genesis: Skipped genesis validation
2023-11-23T16:39:01.095199Z  INFO config: Validating Genesis config and records. This could take a few minutes...
2023-11-23T16:39:01.095523Z  INFO config: All validations have passed!
2023-11-23T16:39:01.095555Z  INFO db_opener: Opening NodeStorage path="/home/ubuntu/.near/data" cold_path="none"
2023-11-23T16:39:01.095589Z  INFO db: Opened a new RocksDB instance. num_instances=1
2023-11-23T16:39:01.114325Z  INFO db: Closed a RocksDB instance. num_instances=0
2023-11-23T16:39:01.114350Z  INFO db_opener: The database exists. path=/home/ubuntu/.near/data
2023-11-23T16:39:01.114361Z  INFO db: Opened a new RocksDB instance. num_instances=1
2023-11-23T16:39:01.273929Z  INFO db: Closed a RocksDB instance. num_instances=0
2023-11-23T16:39:01.273965Z  INFO db: Opened a new RocksDB instance. num_instances=1
2023-11-23T16:39:01.291481Z  INFO db: Closed a RocksDB instance. num_instances=0
2023-11-23T16:39:01.291506Z  INFO db: Opened a new RocksDB instance. num_instances=1
Analysed 43200 blocks between:
Block: height = 106240517, hash = HaqJ1P71gAgHDdMjju72CkZ43uuAS4W3qRVU1LAta4Cq
Block: height = 106196896, hash = AWVbrPJUgy645ZB6eLh3fgZohQ5dQK66RKsh1TrhWBQB

Total gas used: 22407270193906837544

Shard: s0.v1
  Gas usage: 4835423212212028560 (21.6% of total)
  Number of accounts: 197450
  Optimal split:
    split_account: auqurbjawabo.users.kaiching
    gas(account < split_account): 2417713906472325694 (50.0% of shard)
    gas(account >= split_account): 2417709305739702866 (50.0% of shard)

Shard: s1.v1
  Gas usage: 435103165008707214 (1.9% of total)
  Number of accounts: 1
  No optimal split for this shard

Shard: s2.v1
  Gas usage: 6984397849429433430 (31.2% of total)
  Number of accounts: 118485
  Optimal split:
    split_account: kkute9tmsit8.users.kaiching
    gas(account < split_account): 3492199439799673004 (50.0% of shard)
    gas(account >= split_account): 3492198409629760426 (50.0% of shard)

Shard: s3.v1
  Gas usage: 10152345967256668340 (45.3% of total)
  Number of accounts: 81093
  Optimal split:
    split_account: zzzdclzbf6li.users.kaiching
    gas(account < split_account): 5076173908397008194 (50.0% of shard)
    gas(account >= split_account): 5076172058859660146 (50.0% of shard)

10 biggest accounts by gas usage:
#1: token.sweat
    Used gas: 2299587291568145490 (10.3% of total)
#2: earn.kaiching
    Used gas: 2130683133139892087 (9.5% of total)
#3: wallet.kaiching
    Used gas: 612465086458210767 (2.7% of total)
#4: sweat_welcome.near
    Used gas: 470514156151885936 (2.1% of total)
#5: v2.ref-finance.near
    Used gas: 244591219782634243 (1.1% of total)
#6: asset-manager.orderly-network.near
    Used gas: 226419584256414264 (1.0% of total)
#7: aurora
    Used gas: 217558664457482245 (1.0% of total)
#8: app.nearcrowd.near
    Used gas: 207878414604109841 (0.9% of total)
#9: spin.sweat
    Used gas: 174138779272472700 (0.8% of total)
#10: jars.sweat
    Used gas: 128436775874128340 (0.6% of total)
2023-11-23T16:46:26.809965Z  INFO db: Closed a RocksDB instance. num_instances=0

real    7m25.741s
user    5m22.621s
sys     1m4.785s
```

jancionear commented:
> Attention: 315 lines in your changes are missing coverage. Please review.

It'd be nice to test this tool, but for that I would have to mock `ChainStore` and `EpochManager` with some example blockchain data. I will look around to see if there is an easy way to do it.

wacban commented:
> There's a lot of unwraps in the code. Some people dislike that, but in my opinion unwrap() is a good choice for fatal errors in user facing applications. An unwrap() prints the whole error and the exact line where the error has occured, which is great for finding out what went wrong. With Result I often ended up with messages like: Failed: Error: NotFound, which isn't very useful. 

That's fine for tooling in a subcommand. In the usual neard flow we avoid unwraps however and propagate errors up. Personally I also like to log the error at the place where it happens to make it clear what's the place where it occured.

wacban commented:
> The tool isn't very fast - analysing a whole epoch (43200 blocks) takes about 8 minutes. This is most likely caused by the fact that it reads one piece of information from the database at a time. If needed it could probably be sped up by running multiple get_gas_usage_in_block in parallel. 

That's fine, we can run it offline.

wacban commented:
> Codecov Report

Don't worry about testing it, we don't usually write tests for the tools unless we intend them to be used in production. This is just a one off analysis. We can add tests if and when we upgrade it to production code.

wacban commented:
> 10 biggest accounts by gas usage:

Can you make that state per shard? 1% global usage may be way more significant when narrowed down to a shard. Perhaps just print the top 5 per shard.

wacban commented on tools/database/src/block_iterators/height_range.rs:
```diff
@@ -0,0 +1,83 @@
+    type Item = Block;
+
+    fn next(&mut self) -> Option<Block> {
+        if let Some(hash) = self.current_block_hash.take() {
```
mini nit: if you want to make it less nested you could do something like this:
```
let Some(hash) = self.current_block_hash.take() else { return None; }
```
and below return early at each if to totally flatten this method

jancionear commented on tools/database/src/block_iterators/mod.rs:
```diff
@@ -0,0 +1,61 @@
+pub use last_blocks::LastNBlocksIterator;
+
+/// Arguments that user can pass to a command to choose some subset of blocks
+pub struct CommandArgs {
```
I put `CommandArgs` and the logic of creating a block iterator from them in a separate module because I wanted to make it possible to reuse it in another command. In the future we might want to implement another command that analyses a range of blocks, and it would be nice to be able to reuse the existing logic from `analyse-gas-usage`.

wacban replied:
I fully agree, my suggestion was to have separate methods for the different types of iterators. This way you can get rid of the validation of user input from the iterators library. It's only a suggestion and it's just fine as is if you like this more.

jancionear replied:
> my suggestion was to have separate methods for the different types of iterators.

I think this is already provided by the `new` functions of each iterator. You can call `LastNBlocksIterator::new()` or `BlockHeightRangeIterator::new()` to create the desired iterator directly.

jancionear commented on tools/database/src/analyse_gas_usage.rs:
```diff
@@ -0,0 +1,348 @@
+    }
+
+    pub fn get_biggest_accounts(&self) -> impl Iterator<Item = (AccountId, BigGas)> + '_ {
+        self.accounts.iter().rev().map(|(gas, account)| (account.clone(), *gas))
```
It also reverses the range so that the iterator starts at the biggest account instead of the smallest.
Then it also swaps the tuple, I think it looks nicer as `(AccountId, BigGas)` ¯\_(ツ)_/¯

jancionear commented on tools/database/src/analyse_gas_usage.rs:
```diff
@@ -0,0 +1,348 @@
+        }
+        last_analysed_block = Some((block.header().height(), *block.hash()));
+
+        let gas_usage_in_block: GasUsageStats =
```
I personally like to have as many type annotations as possible. In an IDE this doesn't matter because it adds annotations itself, but there are places where this isn't the case - e.g when reviewing the code on github. I feel like having them makes it easier to figure out what's going on - here I would instantly know that we are merging instances of `GasUsageStats`.

wacban replied:
All right, but it does make editing code a bit harder - because then you need to adjust the types on every step as well. Generally speaking the current convention in nearcore is to only add type annotation when needed and we should stick to that.

jancionear replied:
Okay if that's the convention then I will adjust :+1:. I don't really like it, but OK. 

I dislike reading code without type annotations, reading Python is always a pain. One could argue that code is read more often than it's written, so it's more important to make it as readable as possible.

wacban replied:
Sounds like a good topic for zulip or Monday core team discussion :)

wacban commented:
@Ekleog & @nagisa how would you feel about excluding the tools directory from code coverage or setting the target to 0 (for this directory only)? It's looks awfully red for no good reason here ;)

nagisa commented:
Setting the target to 0% seems reasonable enough to me, but I'm not sure I have codecov permissions to make any changes here. Can you please file an issue for this?

wacban commented:
> Setting the target to 0% seems reasonable enough to me, but I'm not sure I have codecov permissions to make any changes here. Can you please file an issue for this?

https://github.com/near/nearcore/issues/10245

jancionear commented on tools/database/src/analyse_gas_usage.rs:
```diff
@@ -0,0 +1,348 @@
+    }
+
+    pub fn add_gas_usage_in_shard(&mut self, shard_uid: ShardUId, shard_usage: GasUsageInShard) {
+        match self.shards.get_mut(&shard_uid) {
```
Using the `enrty()` api it would look like this:
```rust
    pub fn add_gas_usage_in_shard(&mut self, shard_uid: ShardUId, shard_usage: GasUsageInShard) {
        let current_shard_usage =
            self.shards.entry(shard_uid).or_insert_with(|| GasUsageInShard::new());
        current_shard_usage.merge(&shard_usage);
    }
```
I don't really like this version. I feel like the current one is more readable and it avoids copying all elements from one map to the other during `merge`.

jancionear commented on tools/database/src/analyse_gas_usage.rs:
```diff
@@ -0,0 +1,348 @@
+    }
+
+    pub fn add_used_gas(&mut self, account: AccountId, used_gas: BigGas) {
+        let old_used_gas: &BigGas = self.used_gas_per_account.get(&account).unwrap_or(&0);
+        let new_used_gas: BigGas = old_used_gas.checked_add(used_gas).unwrap();
+        self.used_gas_per_account.insert(account, new_used_gas);
```
Initially I didn't go with the entry API because of `checked_add`, but now I tried it and it's actually pretty nice:
```rust
    pub fn add_used_gas(&mut self, account: AccountId, used_gas: BigGas) {
        let account_gas = self.used_gas_per_account.entry(account).or_insert(0);
        *account_gas = account_gas.checked_add(used_gas).unwrap();
    }
```
:+1: changed to entry api

jancionear commented:
> Debugging....

The bug was in `GasUsageInShard::merge()`. By mistake I increased the `used_gas_total` - once when adding accounts from the other instance, and then once again at the end. Because of this the total amount of gas was 2x bigger than it should be, and the logic of shard splitting went crazy.

Fixed in 29190112102f2234420cd548d446799f9b72aa74. Eventually I decided to remove `used_gas_total` altogether. Its presence makes it easier to create bugs, so let's just get rid of it. We can just calculate this information from scratch when needed, it doesn't take that long.

jancionear commented:
v2:
* Fixed the issues mentioned in the review, except for two:
  * One modification to use the entry() api didn't look good to me, I decided not to change it
  * The command is not an enum, I will read up about it
 * Reimplemented `calculate_split()` - it now looks for a split that has the smallest difference between the two halves, and the left half is now inclusive (see commit message for details: f01c80e4e2d5d2cabac15b1bb44ba0a4ab393242)
 * Added unit tests for `calculate_split()`

wacban commented:
I won't be able to review today, sorry. 
@shreyan-gupta Can you have a look in my place?

wacban commented:
> Reimplemented calculate_split() - it now looks for a split that has the smallest difference between the two halves, and the left half is now inclusive (see commit message for details: https://github.com/near/nearcore/commit/f01c80e4e2d5d2cabac15b1bb44ba0a4ab393242) 

& 

>     Left (account <= split_account): 

I don't think this is how it works but I'll double check. Looking at the current shard layout [here](https://github.com/near/nearcore/blob/984d7f927218cfad1a7117a5408ca665fb306bdd/core/primitives/src/shard_layout.rs#L150) shard 1 should contains the single account "aurora" which means that the boundary is the first account of the shard, not the last. For a new boundary aka the split account we want `accounts on the left` < `split account` <= `accounts on the right`.

wacban commented:
> % of shard split half 

Can you make that % of the original shard - without split? The goal is to split the shard as evenly as possible so we want to measure everything relative to the original shard.

wacban commented:
> example output to show how the changes look like 

Thanks, it looks much better now!

wacban commented on tools/database/src/analyse_gas_usage.rs:
```diff
@@ -0,0 +1,348 @@
+type BigGas = u128;
+
+#[derive(Parser)]
+pub(crate) struct AnalyseGasUsageCommand {
```
I found an example in our codebase and it is indeed using a subcommand. If you're curious you can have a look at the StateRootSelector. I don't remember how does it end up working in the cli. Not a big deal, feel free to ignore.

wacban commented on tools/database/src/analyse_gas_usage.rs:
```diff
@@ -0,0 +1,348 @@
+/// A shard can be split into two halves.
+/// This struct represents the result of splitting a shard at `split_account`.
+#[derive(Debug, Clone)]
+struct ShardSplit {
```
Now looking at the updated results, is the split account significant? If so can you print the data in the same format as the tooling for analysing storage? The format is roughly left:middle:right where each is a percentage value e.g. 40%:20%:40%.

jancionear replied:
>  is the split account significant?

I thought that it might be. `split_account` is the first account that made the left half too big and started the right half, so it could be some sort of big account that tipped the scale over.

Looking at the actual results, this is not the case. All `split_accounts` have low gas usage. Because of that I don't know if there is a point in displaying `left:middle:right`, as the middle will just be `0%`.

jancionear replied:
Nevermind, in the new version (where left is defined as <split instead of <=split) the split accounts are actually big. Add the `left:middle:right` display.

jancionear commented:
> For a new boundary aka the split account we want accounts on the left < split account <= accounts on the right.

Damn, you're right.
I looked at `account_id_to_shard_id` and I mistakenly read `boundary_account.cmp(account_id) == Greater` as `account_id > boundary_account`, but that was wrong, it means `account_id < boundary_account`:
```rust
ShardLayout::V1(ShardLayoutV1 { boundary_accounts, .. }) => {
            // Note: As we scale up the number of shards we can consider
            // changing this method to do a binary search rather than linear
            // scan. For the time being, with only 4 shards, this is perfectly fine.
            let mut shard_id: ShardId = 0;
            for boundary_account in boundary_accounts {
                if boundary_account.cmp(account_id) == Greater {
                    break;
                }
                shard_id += 1;
            }
            shard_id
        }
```        

It would be much more readable if it was written using the standard `<` operator :c
```rust
let mut shard_id: ShardId = 0;
for boundary_account in boundary_accounts {
    if account_id < boundary_account {
        break;
    }
    shard_id += 1;
}
```

Thanks! I will fix it.

jancionear commented:
v3:
* Split the shards in a way where the left half is less than split account instead of less or equal.
* Display account gas usage as percent of shard, not shard split
* Show how the gas is distributed in a split `(left, split_account, right)`

Updated the [example output](https://github.com/near/nearcore/pull/10240#issuecomment-1824721481) to show how the latest version looks like.

jancionear commented on tools/database/src/analyse_gas_usage.rs:
```diff
@@ -0,0 +1,585 @@
+                    as_percentage_of(shard_split.gas_right.saturating_sub(split_account_gas), shard_total_gas)
+                );
+                println!("    Left (account < split_account):");
+                let left_accounts =
+                    shard_usage.used_gas_per_account.range(..=shard_split.split_account.clone());
+                display_shard_split_stats(left_accounts, shard_total_gas);
+                println!("    Right (account >= split_account):");
+                let right_accounts =
+                    shard_usage.used_gas_per_account.range(shard_split.split_account..).skip(1);
```
Arght these ranges should also be adjusted

jancionear replied:
done
